### DISCUSSION TITLE
What analog of nuxt 2 render hooks in nuxt3

### DISCUSSION THREAD

**Discussion by JuliWolf** [nuxt/nuxt#20520] [Questions]
Hello. In nuxt2 version we have module based on `render:errorMiddleware` hook
How it is possible to implement the same logic in nuxt3?

Here we catch specific server error and return captcha error page in such cases

Code example
```
const CaptchaHandler: Module = function () {
  this.nuxt.hook('render:errorMiddleware', (app) => {
    app.use((error: CaptchaRpcError, _req, res: Response, next: NextFunction) => {
      if (error.code === 499) {
        const captchaFilePath = path.join(__dirname, '../../client/app/views/captcha.html')

        const captcha = readFileSync(captchaFilePath)
          .toString()
          .replace('{ APP_ERROR_CAPTCHA_PATH }', `${process.env.RPC_URL}${error.data.url}`)

        res.writeHead(499, { 'Content-Type': 'text/html' })
        res.end(captcha)
      }

      next(error)
    })
  })
}
```

**Comment by tibineagu** [comment]
Also interested in this.

We want to migrate the logic enabled by the winston log hook ([here](https://github.com/aaronransley/nuxt-winston-log/blob/f7f4d707b2f2c0ba4dfc6727ebeb0afb383cddd0/index.js#L78C21-L78C43)) so that when we start using nuxt 3 we don't lose any logging logic.

**Reply by TheAlexLichter** [reply]
For this you might listen to Nitro's `error` hook. Example [in my blog post about Nuxt + Sentry](https://www.lichter.io/articles/nuxt3-sentry-recipe/#capturing-errors-with-sentry-in-nitro)

**Comment by TheAlexLichter** [comment]
You could probably use a nitro error handler (e.g. via [`defineNitroErrorHandler`](https://github.com/unjs/nitro/pull/1923)) 🤔
